### DISCUSSION TITLE
Using Crawler to parse HTML Components

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#50416] [Q&A]
I'm trying to use the Crawler component to get data from an HTML component:

```php
$componentHtml = <<<END
    <twig:Alert>
        <twig:block name="footer">
            <button class="btn btn-primary">Claim your prize</button>
        </twig:block>
    </twig:Alert>
END;

 $crawler = new Crawler($componentHtml);

$crawler->filter('twig:Alert')->each(fn(Crawler $node) => dd($node, $node->nodeName(), $componentHtml));

// An exception has been thrown during the rendering of a template ("Pseudo-class "alert" not supported.").

$crawler->filter('twig')->each(fn(Crawler $node) => dd($node, $node->nodeName(), $componentHtml));

// returns a nodeName of html, not the block.

```

I need the raw twig content of the twig blocks within a twig component, perhaps there's an even better way to get it.  I've been using regexs, but I'm guessing there is a way to do this with the DOM Crawler.  Thx.

**Comment by stof** [comment]
If you want to query for namespaced tags in CSS, you need to use `|` as a separator, not `:` (as `:` is the separator for pseudo-classes). However, it would require adding a namespace declaration for the `twig` prefix so that DOMDocument and DOMXPath can work with it. there is `$crawler->registerNamespace()` that can be used for that (you would have to use a fake value for the namespace itself as the Twig component syntax is not using a defined XML namespace, even though the syntax is inspired by the syntax of namespaced nodes).

The root cause of the issue is that what you try to crawl is _not_ HTML. So it does not work out of the box.

**Reply by tacman** [reply]
Thanks, this is a bit of a mystery to me.  I'm hacking away, but still missing something.   How should the namespace be registered, and how should the xpath filter (or regular ->filter) method be called?  In this code snippet, there are no matches.

Or should I be using another approach entirely? 

```php
$componentHtml = <<<END
    <twig:Alert>
        <twig:block name="footer">
            <button class="btn btn-primary">Claim your prize</button>
        </twig:block>
    </twig:Alert>
END;
                    $crawler = new Crawler($componentHtml);
                    $crawler->registerNamespace('twig','fake');
                    foreach (['twig:block', 'twig|block', 'twig', 'block'] as $hack) {
                        $crawler->filterXPath($hack)->each(fn(Crawler $node) => dd($node, $node->nodeName(), $componentHtml));
                    }

                    dd('no matches', $componentHtml);
```
